tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -4354,22 +4355,34 @@ static Value * EvalUnaryIntrinsic(ConstantFP *fpV,
   return Result;
 }
 
-static Value * EvalBinaryIntrinsic(ConstantFP *fpV0, ConstantFP *fpV1,
+static Value * EvalBinaryIntrinsic(Constant *cV0, Constant *cV1,
                                FloatBinaryEvalFuncType floatEvalFunc,
-                               DoubleBinaryEvalFuncType doubleEvalFunc) {
-  llvm::Type *Ty = fpV0->getType();
+                               DoubleBinaryEvalFuncType doubleEvalFunc,
+                               IntBinaryEvalFuncType intEvalFunc) {
+  llvm::Type *Ty = cV0->getType();
   Value *Result = nullptr;
   if (Ty->isDoubleTy()) {
+    ConstantFP *fpV0 = cast<ConstantFP>(cV0);
+    ConstantFP *fpV1 = cast<ConstantFP>(cV1);
     double dV0 = fpV0->getValueAPF().convertToDouble();
     double dV1 = fpV1->getValueAPF().convertToDouble();
     Value *dResult = ConstantFP::get(Ty, doubleEvalFunc(dV0, dV1));
     Result = dResult;
-  } else {
-    DXASSERT_NOMSG(Ty->isFloatTy());
+  } else if (Ty->isFloatTy() || !intEvalFunc) { // maintain previous behavior if no int function is given
+    ConstantFP *fpV0 = cast<ConstantFP>(cV0);
+    ConstantFP *fpV1 = cast<ConstantFP>(cV1);
     float fV0 = fpV0->getValueAPF().convertToFloat();
     float fV1 = fpV1->getValueAPF().convertToFloat();
     Value *dResult = ConstantFP::get(Ty, floatEvalFunc(fV0, fV1));
     Result = dResult;
+  } else {
+    DXASSERT_NOMSG(Ty->isIntegerTy());
```

Comment:
Should add `DXASSERT_NOMSG(intEvalFunc);` here.

Reply:
Or, I suppose neither assert is really necessary, since it will assert/throw on cast if not integer, and crash on call if eval func is nullptr.  They aren't really assumptions that, if violated, would allow code to go off the rails for a while.

Reply:
I personally like having the asserts as documentation of developer intent.  Even when the resulting crash is in the same function, the assert will give a more direct reason.  Although I will agree that this holds true much more for the `isIntegerTy`, than for `intEvalFunc`.  If one tries to invoke it, one obviously assumed it wouldn't be null.

I've added the `DXASSERT_NOMSG(intEvalFunc);` for now, but I'm open to removing it or both if you feel strongly.